mfem/mfem
# ADIOS2 output with ParaView visualization, adios2stream and ADIOS2DataCollection

This PR provides seamless integration of [ADIOS2](https://github.com/ornladios/ADIOS2) output datasets. This PR is ready for review from the mfem community and I'd be happy to address potential feedback and accommodate for the community's preferences.

Overall, this PR provides:

1. The `adios2stream` and `ADIOS2DataCollection` classes for seamless integration of ADIOS2 library outputs [v2.5.0](https://github.com/ornladios/ADIOS2/releases/tag/v2.5.0) similar to the existing mfem's stream abstraction interface.

2. Paraview, as of upcoming [12/10/2019 MPI nightly release](https://www.paraview.org/download/) coverage using the ADIOS2VTXReader is documented under: `config/run_adios2stream.py`

3. Currently the ADIOS2 dependency is supported with mfem CMake-based builds only

To test this PR branch:

1. [Install adios2 v2.5.0](https://adios2.readthedocs.io/en/latest/setting_up/setting_up.html) (macOS users can do `brew install adios2`).

2. Configure mfem with `cmake -DMFEM_USE_ADIOS2=ON -DADIOS2_DIR=/path/to/adios2` and build

3. cd mfem-build/examples copy or soft link the file config/run_adios2stream.py to run current examples with ParaView support. `./run_adios2stream.py ex9p` has transient behavior and would generate the directory `mfem_out_ex9p` which contains the generated adios2 bp directories for each case.

4. Download [ParaView nightly](https://www.paraview.org/download/), go to Version>nightly and select the one from 12/10/2019 or more recent with MPI support.

5. Open any of the bp datasets from 3., select the ADIOSVTXReader and the variable to visualize (e.g. solution) in the variable drop down menu and click on the play button for the transient animation in ex9p. Running in parallel requires server/client mode launching `mpirun -n 4 pvserver`.

Main motivation/long-term goals:
1. Provide flexibility to tackle I/O bottlenecks in MPI environments, examples: set arbitrary number of substreams (N-to-M) where N is MPI processes and 1<M<N. Particularly useful when N is large (1024 or above) to reduce interaction with the filesystem at scale. See #1008 

2. Provide all mfem generated outputs, spatial and temporal data, under a single directory containing binary data (BP format), see #44 . Data interaction can be done through ADIOS2 C++11 [e.g. Python](https://adios2.readthedocs.io/en/latest/api_high/api_high.html#python-high-level-api) APIs, and the command-line [bpls utility](https://adios2.readthedocs.io/en/latest/ecosystem/utilities.html#bpls-inspecting-data).

3. Provide Paraview and VTK out-of-the-box serial and parallel (pvserver) visualization using an automated [schema](https://adios2.readthedocs.io/en/latest/ecosystem/visualization.html#using-vtk-and-paraview). This work is mainly done in VTK on the [VTK::IOADIOS2VTX module reader](https://gitlab.kitware.com/vtk/vtk/tree/master/IO/ADIOS2/VTX).

Current Shortcoming/Limitations:
1. Some AMR cases: ex6p, ex15p
2. Support for mixed elements: (*-mixed meshes)
3. Higher-Order elements (to explore). Currently relying on RefinedGrid.
4. Temporal-varying mesh not tested.
5. Need more tests at scale and miniapps.
6. VisIt support (VisIt is currently using an older VTK version).
7. Currently file-based support only, memory-to-memory insitu requires extra work in VTK


 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1188](https://github.com/mfem/mfem/pull/1188) | @williamfgc | @tzanio | @rcarson3 + @jakubcerveny | 02/16/20 | 03/18/20 | 03/29/20 | |
